src/L2/UpgradeableRegistrarController.sol

```diff
@@ -0,0 +1,686 @@
+    /// @return DiscountDetails associated with the provided `discountKey`.
+    function discounts(bytes32 discountKey) external view returns (DiscountDetails memory) {
+        return _getURCStorage().discounts[discountKey];
+    }
+
+    /// @notice Fetches the payment receiver from storage.abi
+    ///
+    /// @return The address of the payment receiver.
+    function paymentReceiver() external view returns (address) {
+        return _getURCStorage().paymentReceiver;
+    }
+
+    /// @notice Fetches the price oracle from storage.
+    ///
+    /// @return The stored prices oracle.
+    function prices() external view returns (IPriceOracle) {
+        return _getURCStorage().prices;
+    }
+
+    /// @notice Fetches the Reverse Registrar from storage.
+    ///
+    /// @return The stored Reverse Registrar.
+    function reverseRegistrar() external view returns (IReverseRegistrarV2) {
+        return _getURCStorage().reverseRegistrar;
+    }
+
+    /// @notice Checks the rent price for a provided `name` and `duration`.
+    ///
+    /// @param name The name to check the rent price of.
+    /// @param duration The time that the name would be rented.
+    ///
+    /// @return price The `Price` tuple containing the base and premium prices respectively, denominated in wei.
+    function rentPrice(string memory name, uint256 duration) public view returns (IPriceOracle.Price memory price) {
+        bytes32 label = keccak256(bytes(name));
+        price = _getURCStorage().prices.price(name, _getExpiry(uint256(label)), duration);
+    }
+
+    /// @notice Checks the register price for a provided `name` and `duration`.
+    ///
+    /// @param name The name to check the register price of.
+    /// @param duration The time that the name would be registered.
+    ///
+    /// @return The all-in price for the name registration, denominated in wei.
+    function registerPrice(string memory name, uint256 duration) public view returns (uint256) {
+        IPriceOracle.Price memory price = rentPrice(name, duration);
+        return price.base + price.premium;
+    }
+
+    /// @notice Checks the discounted register price for a provided `name`, `duration` and `discountKey`.
+    ///
+    /// @dev The associated `DiscountDetails.discount` is subtracted from the price returned by calling `registerPrice()`.
+    ///
+    /// @param name The name to check the discounted register price of.
+    /// @param duration The time that the name would be registered.
+    /// @param discountKey The uuid of the discount to apply.
+    ///
+    /// @return price The all-in price for the discounted name registration, denominated in wei. Returns 0
+    ///         if the price of the discount exceeds the nominal registration fee.
+    function discountedRegisterPrice(string memory name, uint256 duration, bytes32 discountKey)
+        public
+        view
+        returns (uint256 price)
+    {
+        uint256 discount = _getURCStorage().discounts[discountKey].discount;
```

Comment:
just an observation but seems interesting to me that discounts are fixed and independent of the `registerPrice` for the name v.s. a percentage discounted etc.